### PR TITLE
Expert filter properties

### PR DESCRIPTION
**Pull request by souissimai**

### CODE REVIEW

File: src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java [hunk 1]
```diff
@@ -62,63 +65,99 @@ public static AbstractExpertRule entityToDto(ExpertRuleEntity filterEntity) {
...
+                    return BooleanExpertRule.builder()
+                            .field(booleanFilterEntity.getField())
+                            .operator(booleanFilterEntity.getOperator())
+                            .value(Boolean.parseBoolean(booleanFilterEntity.getValue()))
+                            .build();
+                } else {
+                    throw new PowsyblException("Invalid entity type for BOOLEAN rule");
+                }
+
             }
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java:78]: In my opinion, we can use casting directly, this situation does normally never occur.. idem for others

File: src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java [hunk 2]
```diff
@@ -148,15 +187,43 @@ public ExpertFilterEntity fromDto(AbstractFilter dto) {
...
+                    .id(UUID.randomUUID())
+                    .combinator(filter.getCombinator())
+                    .operator(filter.getOperator())
+                    .dataType(filter.getDataType())
+                    .field(filter.getField())
+                    .value(filter.getStringValue());
+        }
+        if (expertRuleEntityBuilder == null) {
+            throw new PowsyblException("Unsupported data type: " + filter.getDataType());
+        }
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java:221]: In my opinion, we can extract to a method: getBuilder(rule) which return a builder in order to reuse in dtoToEntity and in the map() in dtoToEntities.

    public static ExpertRuleEntity.ExpertRuleEntityBuilder<?, ?> getBuilder(AbstractExpertRule filter) {
        ExpertRuleEntity.ExpertRuleEntityBuilder<?, ?> expertRuleEntityBuilder = null;
        if (filter.getDataType() == DataType.COMBINATOR) {
            expertRuleEntityBuilder = ExpertRuleEntity.builder()
                    .id(UUID.randomUUID())
                    .combinator(filter.getCombinator())
                    .operator(filter.getOperator())
                    .dataType(filter.getDataType())
                    .field(filter.getField());
        }
        if (filter.getDataType() == DataType.PROPERTIES) {
            PropertiesExpertRule propertiesRule = (PropertiesExpertRule) filter;
            expertRuleEntityBuilder = ExpertRulePropertiesEntity.builder()
                    .id(UUID.randomUUID())
                    .combinator(filter.getCombinator())
                    .operator(filter.getOperator())
                    .dataType(filter.getDataType())
                    .field(filter.getField())
                    .propertyValues(propertiesRule.getPropertyValues())
                    .propertyName(propertiesRule.getPropertyName());

        } else if (filter.getDataType() == DataType.BOOLEAN || filter.getDataType() == DataType.NUMBER || filter.getDataType() == DataType.STRING || filter.getDataType() == DataType.ENUM || filter.getDataType() == DataType.FILTER_UUID) {
            expertRuleEntityBuilder = ExpertRuleValueEntity.builder()
                    .id(UUID.randomUUID())
                    .combinator(filter.getCombinator())
                    .operator(filter.getOperator())
                    .dataType(filter.getDataType())
                    .field(filter.getField())
                    .value(filter.getStringValue());
        }
        if (expertRuleEntityBuilder == null) {
            throw new PowsyblException("Unsupported data type: " + filter.getDataType());
        }

        return expertRuleEntityBuilder;
    }

That can avoid duplication of code

File: src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java [hunk 3]
```diff
@@ -166,14 +233,27 @@ private static List<ExpertRuleEntity> dtoToEntities(List<AbstractExpertRule> rul
...
+                    } else {
+                        expertRuleEntityBuilder = ExpertRuleValueEntity.builder()
+                                .id(UUID.randomUUID())
+                                .combinator(rule.getCombinator())
+                                .operator(rule.getOperator())
+                                .dataType(rule.getDataType())
+                                .field(rule.getField())
+                                .value(rule.getStringValue())
+                                .parentRule(parentRuleEntity);
+                    }
```
Comment (thangqp) [inline src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java:256]: = getBuilder(rule).parentRule(parentRuleEntity)

**Review by thangqp — APPROVED** [review]

File: src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java [hunk 4]
```diff
@@ -62,63 +65,99 @@ public static AbstractExpertRule entityToDto(ExpertRuleEntity filterEntity) {
...
+                            .operator(propertiesFilterEntity.getOperator())
+                            .propertyValues(propertiesFilterEntity.getPropertyValues())
+                            .propertyName(propertiesFilterEntity.getPropertyName())
+                            .build();
+                } else {
+                    throw new PowsyblException("Invalid entity type for PROPERTIES rule");
                 }
-                return ruleBuilder.build();
             }
             default -> throw new PowsyblException("Unknown rule data type: " + filterEntity.getDataType() + ", supported data types are: COMBINATOR, BOOLEAN, NUMBER, STRING, ENUM");
```
[the remark this answers is not in the document]
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java]: done

File: src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java [hunk 5]
```diff
@@ -148,15 +187,43 @@ public ExpertFilterEntity fromDto(AbstractFilter dto) {
...
+                    .id(UUID.randomUUID())
+                    .combinator(filter.getCombinator())
+                    .operator(filter.getOperator())
+                    .dataType(filter.getDataType())
+                    .field(filter.getField())
+                    .propertyValues(propertiesRule.getPropertyValues())
+                    .propertyName(propertiesRule.getPropertyName());
+
+        } else if (filter.getDataType() == DataType.BOOLEAN || filter.getDataType() == DataType.NUMBER || filter.getDataType() == DataType.STRING || filter.getDataType() == DataType.ENUM || filter.getDataType() == DataType.FILTER_UUID) {
+            expertRuleEntityBuilder = ExpertRuleValueEntity.builder()
```
[the remark this answers is not in the document]
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java]: done

File: src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java (hunk 1 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java]: done

File: src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java (hunk 2 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java]: done

File: src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java (hunk 3 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/filter/server/repositories/proxies/expertfiler/ExpertFilterRepositoryProxy.java]: done

**Review by ayolab — APPROVED** [review]

**Pull request merged**